### PR TITLE
Add PreferenceStore interface

### PR DESCRIPTION
**Pull request by cdupuis**

### CODE REVIEW

**Comment by cdupuis** [comment]
I think I can improve on backwards compatibility here. Two things that are of concern:

* make `preferenceFactory` optional on the config options
* `chooseAndSetGoals` should not make passing over the `PreferenceStoreFactory` required. It can fall back to the `NoPreferenceStore` implementation which should be fine. If they want to use preferences from within `PushTests`, they have to pass the proper store. But this would be backwards compatible.

**Comment by ddgenome** [comment]
If it affects users, I think we should try to make it backward compatible. If it does not affect users, I think we should make the implementation as usable as possible.

File: lib/api/context/SdmContext.ts [hunk 1]
```diff
@@ -48,6 +49,10 @@ export interface SdmContext {
      */
     credentials: ProjectOperationCredentials;
 
+    /**
+     *
```
Comment (ddgenome) [inline lib/api/context/SdmContext.ts]: Did you mean to put some sort of comment in here?

Comment (cdupuis) [reply lib/api/context/SdmContext.ts]: I think i should, shouldn't I?

**Pull request merged**